Global style revisions: redesign style revision items

ramonjd opened the pull request:
## What?

Resolves https://github.com/WordPress/gutenberg/issues/55647

This PR:

- shows humantimediff for changes < 1 day. Other dates are displayed as longer formats
- redesigns timeline and revision elements


https://github.com/WordPress/gutenberg/assets/6458278/6c59ef5e-d347-4c0c-abec-b179b672e23e


## Testing Instructions
Now open the styles revisions panel.

Check that the revision items work and the active hover states are correct.

ramonjd committed:
First commit:
- add basic function to return translated labels of top-level changes
- showing humantime diff for changes < 1 day
- rearranging elements

ramonjd committed:
- timeline redesign
- minor optimization refactoring

ramonjd committed:
There is no "colors" property

ramonjd committed:
Some style tweaks

ramonjd committed:
Tweak font sizes

ramonjd commented on packages/edit-site/src/components/global-styles/screen-revisions/revisions-buttons.js:
```diff
@@ -65,25 +72,40 @@ function getRevisionLabel( revision ) {
+				const displayDate =
+					modified &&
+					dateNowInMs - modifiedDate.getTime() > DAY_IN_MILLISECONDS
+						? dateI18n( dateFormat, modifiedDate )
```
The date is formatted as **OCTOBER 20 2023**

This doesn't match the designs, which specify an abbreviated month — **OCT 20 2023** — however I think it's important to introduce new formats using the date library and into Core so that themes can override them.

So TL;DR I can open a new PR and corresponding core patch to introduce the abbreviated month format if folks think it's worth it.

ramonjd committed:
Display Unsaved for unsaved changes

apeatling reviewed (commented):
I've noticed a bug where the previous revisions changes are added to the next revision. It's like the list isn't getting cleared. Try changing a couple of distinct things (other than typography), then save. Then change a global typography setting and save. You'll notice that the most recent revision lists not just typography changes.

In this example I only changed typography in the latest revision:

<img width="277" alt="Screenshot 2023-11-16 at 1 04 14 PM" src="https://github.com/WordPress/gutenberg/assets/1464705/c61f50f3-c4f0-4819-9c52-bf91113d86a7">

apeatling commented:
I also think we need to adjust the line height of the changes in the list to match everything else. You can see the difference between the line height in the description paragraph at the top, and the items in the list:

<img width="351" alt="Screenshot 2023-11-16 at 1 06 49 PM" src="https://github.com/WordPress/gutenberg/assets/1464705/a3c09b94-5f3d-4894-8eb0-271143bc8f55">

apeatling commented on packages/edit-site/src/components/global-styles/screen-revisions/get-revision-changes.js:
```diff
@@ -0,0 +1,40 @@
+import { __ } from '@wordpress/i18n';
+
+const translationMap = {
+	blocks: __( 'Block styles' ),
```
Can we break out block names in a follow up? For example show "Paragraph block typography" etc instead of a generic "Block styles". I think block styles are going to change a bunch and a lot of "Block styles" being listed is not super useful.

ramonjd replied:
For blocks it makes sense. I think we have to put a limit on it at some point as it needs to be calculated for every revision item on render.

I'll see how memoizing/caching each one might work too. 

@c4rl0sbr4v0 explored comparing the style/settings tree deeply: https://github.com/WordPress/gutenberg/pull/52829#issuecomment-1659027089

ramonjd commented:
> I've noticed a bug where the previous revisions changes are added to the next revision. It's like the list isn't getting cleared.

Thanks for testing. It's a naive check in that it only describes an individual revision's changes. It doesn't do a comparison with the previous revision, which it probably should.

I shied away from that approach for performance reasons. If we keep the comparison to the very top level it should be okay. 

I'll have a look.

ramonjd commented:
> It doesn't do a comparison with the previous revision, which it probably should.

Actually, now that I reread that, we'd probably need to do a deep comparison to know what's changed, even at the top level.

So for example if a revision makes changes to `styles.spacing.margin.marginLeft`, that potentially 3-levels deep for every style.

🤔

ramonjd committed:
Remove revision changes labels. To be done in a follow up

ramonjd commented:
@apeatling I've been testing various methods to get some kind of overview of changes.

I think it's best for a follow up PR because it would contain a few changes and a lot of logic that would be good to have reviewed separately. For now I've removed the relevant code.

Here are some initial thoughts for the follow up, assuming the goal of rendering a meaningful description of changes. Something like `"Typography, background color, color palette, list block typography, link color..."`

- a changeset between revisions could be very large. 
    - How many results should we display? And which ones?
    - How granular should it be? For example, should we itemize all block.typography properties? 
    - Which results should we exclude? Do we include all settings such as font changes no matter how repetitive?
- performance needs to be measured. Taking into account settings such as fonts and colors, the object nesting is very deep. We can limit the depth and cache comparisons by revision id to help performance. We could also only show a list of changes when an item is selected, so "expand" the revision item somehow.
- How to maintain a list of translations for `n` styles?

I've taken @c4rl0sbr4v0's function from https://github.com/WordPress/gutenberg/pull/52829/files#diff-6baeb967c4bc1836222ed3912fb25cfdba460429895110e58ee23413e35498c6R97 and changed it a little:

- filtering the results by `hasChanged` to only return changed values from the previous revision.
- randomizing the order, then slicing the array so as not to return too many results, and not the same one between revisions.
- caching each result under a revision id
- returning early when max depth is reached.

It works okay, but needs a lot of refinement particularly surrounding which changes to show. 

Switching between style variation for example creates 100 changes, many of them settings changes. This makes for a lot of repetitiveness.

So I believe we need to think and experiment a bit in order to balance all the concerns.

andrewserong commented:
Just catching up on the discussion — good idea skipping the complexity of comparing the changes of each revision for now, it does sound like a good thing to address separately and explore our options.

One idea that came to me from reading the discussion: this might be ridiculous, but would an option be to potentially store a message in the `post_excerpt` field when a revision is saved? That way a record of the changes that were made is effectively cached at the time that the change was made. It could also have the benefit of being user-editable. Either at the time of saving or afterwards, a user could go in and attach a message to a particular global styles revisions — like, "Updated styles for Halloween" or "Implemented brand refresh"... essentially folks could use the field like a commit message. Just an idea, though, as it could be a can of worms (i.e. the message is in the db so can't be changed via code)!

andrewserong commented on packages/edit-site/src/components/global-styles/screen-revisions/revisions-buttons.js:
```diff
@@ -12,33 +12,35 @@ import { dateI18n, getDate, humanTimeDiff, getSettings } from '@wordpress/date';
 import { store as coreStore } from '@wordpress/core-data';
 import { useSelect } from '@wordpress/data';
 
+const DAY_IN_MILLISECONDS = 60 * 60 * 1000 * 24;
+const dateFormat = getSettings().formats.date;
+const datetimeAbbreviatedFormat = getSettings().formats.datetimeAbbreviated;
+
```
Is it safe to call `getSettings()` outside of a render context? This means that `getSettings()` is called when the file is imported, rather than when anything within the file is executed. In the `LatestPostsEdit` component, for example, it's called just before render here: https://github.com/WordPress/gutenberg/blob/2d86e9053e79e7d2ab37e00c24ac7145980ad959/packages/block-library/src/latest-posts/edit.js#L431

My understanding is that `wp.data.setSettings()` will get called at a potentially indeterminate time, so it's likely safer to get settings during render? I think this is where we're calling `setSettings`: https://github.com/WordPress/gutenberg/blob/6752a2a8d418a8d1231e3ddc498d49ca4c7ed6c6/lib/compat/wordpress-6.4/script-loader.php#L33

ramonjd replied:
Good point. 

The hook that renders `wp.data.setSettings()` to the HTML (`wp_default_scripts`) fires pretty early on, so in theory it should be in the HTML before the JS is finished parsing all that react code 😄 

But it wouldn't hurt to move it as you said. 

Thanks!!

andrewserong commented on packages/edit-site/src/components/global-styles/screen-revisions/revisions-buttons.js:
```diff
@@ -12,33 +12,35 @@ import { dateI18n, getDate, humanTimeDiff, getSettings } from '@wordpress/date';
-	const authorDisplayName = revision?.author?.name || __( 'User' );
-
-	if ( 'parent' === revision?.id ) {
+function getRevisionLabel( id, isLatest, authorDisplayName, modifiedDate ) {
```
Tiny nit: would it be worth switching over to an object as a param so that it's easier to add or remove elements for the label rather than use positional params?

ramonjd replied:
Nothing against that approach. The function is internal to the component so I'm not sure of the value aside from making refactoring easier later.

andrewserong replied:
Feel free to leave it if you like, since it's not exposed anywhere 🙂

andrewserong commented on packages/edit-site/src/components/global-styles/screen-revisions/style.scss:
```diff
@@ -8,62 +8,80 @@
+		left: $grid-unit-20;
+		top: 0;
+		width: 0;
+		border: 0.5px solid #ddd;
```
Tiny nit: is it worth using `$gray-300` rather than `#ddd` specifically?

ramonjd replied:
It is worth it!

andrewserong commented on packages/edit-site/src/components/global-styles/screen-revisions/style.scss:
```diff
@@ -78,20 +96,26 @@
+	.edit-site-global-styles-screen-revisions__date {
+		text-transform: uppercase;
+		font-weight: 600;
+		font-size: 12px;
```
For each of the `font-size: 12px` rules, should we reference `$helptext-font-size` instead?

ramonjd replied:
I saw that but wasn't sure whether I should use it as it's specific to help text and might change?

We could add another var I guess 🤔

andrewserong replied:
Feel free to leave it, then, it's only two hard-coded lines 🙂

andrewserong reviewed (commented):
This is looking great to me in testing @ramonjd! Love the level of polish for the beginning and ends of the list so that the line never extends beyond the dot:

### Beginning

<img width="279" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/75760c07-9b1b-4ad4-bd70-bf2f5dbdc831">

### End

<img width="278" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/ec730a37-c08c-489d-a271-fde35dcc18db">

Just left a few comments, mostly just tiny nits. The main thing I was wondering about was making sure we call `getSettings()` at render time so that if the date settings have been updated anywhere else, they're reflected in time.

Not for this PR, but looking at the repeated dates when many changes occurred on the one day makes we wonder if we should expose the time somewhere, too (or explore group-by-date collapsing), otherwise it can look a bit repetitive:

<img width="279" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/b558ae86-5eec-4c9c-aae9-4baeab4b0c9f">

Not a blocker for now, though 🙂

ramonjd commented:
Thanks for looking at this!

> would an option be to potentially store a message in the post_excerpt field when a revision is saved? 

It's an interesting idea!

`post_excerpt` doesn't exist on the global styles post type (it doesn't inherit `WP_Post_Controller`) so we'd have to activate that field for both the post type and the revisions. That's not such a huge deal.

It did occur to me to move the work to the backend, though many of the questions in the above [comment](https://github.com/WordPress/gutenberg/pull/55913#issuecomment-1815609449) would still need to be answered.

That is, I see the main challenge to be how and when to compile the data, and what that data would look like. 

If we saved an excerpt, we could avoid expensive object comparison by hooking into the global styles provider in the site editor and keep an audit of currently unsaved settings/style changes. Using that we could send a version of that data to the database when updating the global styles post.

The catch is that any updates done over the API elsewhere might not compile such a change set, so it probably has to be done on the backend, in which case we push the performance burden to the server. Perhaps a dedicated endpoint to grab a change set might be more appropriate.

To get back to the feature at a higher level, I had this question in my mind: What do we need to communicate in these revisions change summaries?

In my view, if we're just dealing with copy, they should:

- display changes to root and block styles, different from the previous revision
- display a max of `n` changes with out going too deep or being too repetitive. I'd therefore imagine changes to prioritize changes at the block support group level e.g., `Global letter spacing, global padding, group block border, color palettes and more.` Contrast this with block support item level: `Column block border left color`. This then risks repetitiveness if several revisions update Column block borders.

It's tricky! But if they're all reasonable assumptions, and we don't yet have a use case for a full change set (like in this [design proposal](https://github.com/WordPress/gutenberg/pull/52829#issuecomment-1648177092)), then I still think we could trial it clientside with the appropriate caching and failsafes to avoid going too deep into the tree.

I'd be open to think about different UXs here too. E.g., clicking on the item to grab a changeset from the server and rendering the result.

🤷🏻

andrewserong commented:
> But if they're all reasonable assumptions, and we don't yet have a use case for a full change set (like in this https://github.com/WordPress/gutenberg/pull/52829#issuecomment-1648177092), then I still think we could trial it clientside with the appropriate caching and failsafes to avoid going too deep into the tree.

Agreed. Thanks for thinking this through in detail! I think a minimal clientside approach would be great to test out the ideas before baking anything in server-side 👍

ramonjd committed:
Moving date format setting call into the comoponent
rejigging getLabel function
adding css var

ramonjd commented:
>  we should expose the time somewhere, too (or explore group-by-date collapsing), otherwise it can look a bit repetitive

Great idea!! I'll try it out. 

Agree that it looks a bit samey 😄

andrewserong reviewed (approved):
Thanks for the updates! This is looking great to me, I reckon it looks good to merge, and we can continue exploring the other ideas in follow-ups.

LGTM! ✨

mirka commented on packages/edit-site/src/components/global-styles/screen-revisions/style.scss:
```diff
@@ -8,62 +8,80 @@
 	}
 }
 
 .is-selected {
+	color: var(--wp-components-color-accent, var(--wp-admin-theme-color, #007cba));
+	background: rgba(var(--wp-admin-theme-color--rgb), 0.04);
```
These color and background styles for the `.is-selected` modifier is not scoped to a specific `.edit-site-global-styles-screen-revisions__` class, so it's affecting every component that uses a `is-selected` modifier. For example, the `PaletteEdit` in the Global Styles color palette editor:

<img width="274" alt="Affected PaletteEdit" src="https://github.com/WordPress/gutenberg/assets/555336/744d24b0-0fd2-491b-8a8b-0b203100615d">

@ramonjd I don't have much context on these Revisions UIs — do you have bandwidth to scope this CSS properly to the relevant classes?

andrewserong replied:
Oh, great catch! I can look into this.

andrewserong replied:
Fix here: https://github.com/WordPress/gutenberg/pull/58228

ramonjd replied:
🤦🏻  Thanks for catching @mirka and for the quick fix @andrewserong
